File: sdks/typescript/src/storage.ts
/**
 * Default in-memory receipt store. Production deployments should pass a
 * persistent ReceiptStore (e.g. SQLite, Postgres) via ServerConfig.storage.
 */

import type { Incident, Receipt, ReceiptStore, Subscription } from './types.js';

export class MemoryStore implements ReceiptStore {
  private receipts: Receipt[] = [];
  private subscriptions = new Map<string, Subscription>();
  private incidents: Incident[] = [];
  private chainTip = 'genesis';
  private buckets = new Map<string, { count: number; expiresAt: number }>();

  insertReceipt(r: Receipt): void {
    this.receipts.push(r);
  }

  receiptsByPrincipal(principal: string, limit: number): Receipt[] {
    return this.receipts
      .filter((r) => r.principal_did === principal)
      .slice(-limit)
      .reverse();
  }

  allReceipts(limit: number): Receipt[] {
    return this.receipts.slice(-limit).reverse();
  }

  deleteByPrincipal(principal: string): number {
    const before = this.receipts.length;
    this.receipts = this.receipts.filter((r) => r.principal_did !== principal);
    return before - this.receipts.length;
  }

  getChainTip(): string {
    return this.chainTip;
  }

  setChainTip(hash: string): void {
    this.chainTip = hash;
  }

  insertSubscription(s: Subscription): void {
    this.subscriptions.set(s.subscription_id, s);
  }

  cancelSubscription(id: string): boolean {
    const s = this.subscriptions.get(id);
    if (!s) return false;
    s.status = 'canceled';
    s.canceled_at = new Date().toISOString();
    return true;
  }

  activeSubscription(principal: string): Subscription | null {
    for (const s of this.subscriptions.values()) {
      if (s.principal_did === principal && s.status === 'active') return s;
    }
    return null;
  }

  insertIncident(i: Incident): void {
    this.incidents.push(i);
  }

  listIncidents(limit: number): Incident[] {
    return this.incidents.slice(-limit).reverse();
  }

  rateBucketIncrement(key: string, ttlMs: number): number {
    const now = Date.now();
    const existing = this.buckets.get(key);
    if (existing && existing.expiresAt > now) {
      existing.count += 1;
      return existing.count;
    }
    this.buckets.set(key, { count: 1, expiresAt: now + ttlMs });
    if (Math.random() < 0.01) {
      for (const [k, v] of this.buckets) {
        if (v.expiresAt <= now) this.buckets.delete(k);
      }
    }
    return 1;
  }
}

File: sdks/python/src/oap_sdk/storage.py
"""Default in-memory ReceiptStore."""

from __future__ import annotations

import random
import time
from typing import Any


class MemoryStore:
    def __init__(self) -> None:
        self._receipts: list[dict[str, Any]] = []
        self._subscriptions: dict[str, dict[str, Any]] = {}
        self._incidents: list[dict[str, Any]] = []
        self._chain_tip = "genesis"
        self._buckets: dict[str, dict[str, float]] = {}

    def insert_receipt(self, receipt: dict[str, Any]) -> None:
        self._receipts.append(receipt)

    def receipts_by_principal(self, principal: str, limit: int) -> list[dict[str, Any]]:
        filtered = [r for r in self._receipts if r.get("principal_did") == principal]
        return list(reversed(filtered[-limit:]))

    def all_receipts(self, limit: int) -> list[dict[str, Any]]:
        return list(reversed(self._receipts[-limit:]))

    def delete_by_principal(self, principal: str) -> int:
        before = len(self._receipts)
        self._receipts = [r for r in self._receipts if r.get("principal_did") != principal]
        return before - len(self._receipts)

    def get_chain_tip(self) -> str:
        return self._chain_tip

    def set_chain_tip(self, h: str) -> None:
        self._chain_tip = h

    def insert_subscription(self, sub: dict[str, Any]) -> None:
        self._subscriptions[sub["subscription_id"]] = sub

    def cancel_subscription(self, sub_id: str) -> bool:
        sub = self._subscriptions.get(sub_id)
        if not sub:
            return False
        sub["status"] = "canceled"
        sub["canceled_at"] = _now_iso()
        return True

    def active_subscription(self, principal: str) -> dict[str, Any] | None:
        for sub in self._subscriptions.values():
            if sub.get("principal_did") == principal and sub.get("status") == "active":
                return sub
        return None

    def insert_incident(self, incident: dict[str, Any]) -> None:
        self._incidents.append(incident)

    def list_incidents(self, limit: int) -> list[dict[str, Any]]:
        return list(reversed(self._incidents[-limit:]))

    def rate_bucket_increment(self, key: str, ttl_ms: int) -> int:
        now = time.time() * 1000
        existing = self._buckets.get(key)
        if existing and existing["expires_at"] > now:
            existing["count"] += 1
            return int(existing["count"])
        self._buckets[key] = {"count": 1, "expires_at": now + ttl_ms}
        if random.random() < 0.01:
            self._buckets = {
                k: v for k, v in self._buckets.items() if v["expires_at"] > now
            }
        return 1


def _now_iso() -> str:
    from datetime import datetime, timezone

    return datetime.now(timezone.utc).isoformat().replace("+00:00", "Z")
